xarray/core/weighted.py

```diff
@@ -100,12 +100,16 @@ def __init__(self, obj, weights):
         if not isinstance(weights, DataArray):
             raise ValueError("`weights` must be a DataArray")
 
-        if weights.isnull().any():
-            raise ValueError(
-                "`weights` cannot contain missing values. "
-                "Missing values can be replaced by `weights.fillna(0)`."
-            )
-
+        def _weight_check(w):
+            if w.isnull().any():
+                raise ValueError(
+                    "`weights` cannot contain missing values. "
+                    "Missing values can be replaced by `weights.fillna(0)`."
+                )
+            else:
+                return w
+
+        weights = weights.map_blocks(_weight_check, template=weights)
```

Comment:
```suggestion
        weights = dask.array.map_blocks(_weight_check.data)
```

should be more efficient.

Reply:
This seems to cause a bunch of problems for me. Could you explain why this is more efficient? I might be missing something due to my lack of experience with these internals.

Reply:
`weights.map_blocks` wraps `dask.array.map_blocks` and does a bunch of extra stuff. So calling `dask.array.map_blocks` is more efficient.